presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushProjectionIntoTableScan.java

```diff
@@ -0,0 +1,130 @@
+
+    private final Metadata metadata;
+    private final TypeAnalyzer typeAnalyzer;
+
+    public PushProjectionIntoTableScan(Metadata metadata, TypeAnalyzer typeAnalyzer)
+    {
+        this.metadata = metadata;
+        this.typeAnalyzer = typeAnalyzer;
+    }
+
+    @Override
+    public Pattern<ProjectNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(ProjectNode project, Captures captures, Context context)
+    {
+        TableScanNode tableScan = captures.get(TABLE_SCAN);
+
+        List<ConnectorExpression> projections;
+        try {
+            projections = project.getAssignments()
+                    .getExpressions().stream()
+                    .map(expression -> ConnectorExpressionTranslator.translate(
+                            context.getSession(),
+                            expression,
+                            typeAnalyzer,
+                            context.getSymbolAllocator().getTypes()
+                    ))
+                    .collect(Collectors.toList());
+        }
+        catch (UnsupportedOperationException e) {
+            // some expression not supported by translator, skip
+            return Result.empty();
+        }
+
+        Map<String, ColumnHandle> assignments = tableScan.getAssignments()
+                .entrySet().stream()
+                .collect(Collectors.toMap(entry -> entry.getKey().getName(), Map.Entry::getValue));
+
+        Optional<ProjectionApplicationResult<TableHandle>> result = metadata.applyProjection(context.getSession(), tableScan.getTable(), projections, assignments);
+        if (!result.isPresent()) {
+            return Result.empty();
+        }
+
+        List<Symbol> newScanOutputs = new ArrayList<>();
+        Map<Symbol, ColumnHandle> newScanAssignments = new HashMap<>();
+        Map<String, Symbol> variableMappings = new HashMap<>();
+        for (ProjectionApplicationResult.Assignment assignment : result.get().getAssignments()) {
+            Symbol symbol = context.getSymbolAllocator().newSymbol(assignment.getVariable(), assignment.getType());
+
+            newScanOutputs.add(symbol);
+            newScanAssignments.put(symbol, assignment.getColumn());
+            variableMappings.put(assignment.getVariable(), symbol);
+        }
+
+        // TODO: ensure newProjections.size == original projections.size
+
+        List<Expression> newProjections = result.get().getNewProjections().stream()
+                .map(expression -> ConnectorExpressionTranslator.translate(expression, variableMappings))
+                .collect(Collectors.toList());
+
+        Assignments.Builder newProjectionAssignments = Assignments.builder();
+        for (int i = 0; i < project.getOutputSymbols().size(); i++) {
+            newProjectionAssignments.put(project.getOutputSymbols().get(i), newProjections.get(i));
+        }
+
+        return Result.ofPlanNode(
+                new ProjectNode(
```

Comment:
Don't you need to have yet another projection that would map new produced symbols to the origin one (from project node before the rewrite)?

Reply:
No, that's taken care of by making the projection outputs be the same as the original project:
```
newProjectionAssignments.put(project.getOutputSymbols().get(i), newProjections.get(i));
```